### ISSUE TITLE
Records in files are sometimes skipped

### ISSUE THREAD

**Issue by andrew-steinkopf** [MistakenIdentity/WillPower#1]
I am trying to process a few different files and in certain cases, records are skipped. For example, in one instance I have a file with 50 records. In most cases, all 50 records are read but sometimes I only get 48 or 49 records.

In another situation, I have a file with 3683 records but the records the Parser has is never consistent and it is never 3683. I have 3674, 3673, 3669, and 3675. These were on subsequent runs. I did start getting the same numbers (3675 and 3673 again) after that but it is wildly inconsistent in terms of how many records it reads. And it never seems to get to the "right" number of records meaning important data is being lost.

**Comment by DarkAngel766** [comment]
Hi Andrew.  
Did you happen to look in the Exceptions collections in the Records after parsing?
I'm assuming this is a 2d file. Are you reading this from a file or a stream/byte[]?
Are there any samples you can provide?

Kind Regards,

Willow

**Comment by andrew-steinkopf** [comment]
I did but the problem seems to be that the records simply don't exist so there are no exceptions for those records. Of the records that exist, none have exceptions. It seems like there needs to be a top level exceptions object above the records to help explain why certain ones didn't parse. Unless there is and I am missing it?

I tried from a file and from a stream and both seem to have the same problem.

As for the 2d file My code is built similar to your 3d file example as there are a few different file types and that is what seemed to make sense. I may not completely understand the difference between 2d and 3d though. So maybe this is where part of the issue resides.

Unfortunately, I cannot provide examples as the files I am working with are tax records and have PII in them that I wouldn't be at liberty to share.

**Comment by DarkAngel766** [comment]
I don't suppose you could create a sample file using the write feature?  I will take your suggestion on a top level exceptions collection to heart and make that happen, but I really need more to go on.

Please note, I do not have access to an AS/400 or similar system anymore to produce native files.

**Comment by andrew-steinkopf** [comment]
I can give it a shot to try to create some similar files with the private info stripped out. While the most recent file I mentioned had Comp-3 fields, I probably will create a file without them since other files had a similar problem so I know it isn't the comp-3 fields. Or, I think it isn't. Might be a couple days depending on how difficult it ends up being to strip out that info.

Also, does this code support variable length records? I noticed I put in a length in the IFileLayout, but I believe I have two files types with variable record lengths. Just want to see if that is supported in any way.

**Comment by andrew-steinkopf** [comment]
I couldn't figure out how to write out a file using your parser so I just found another way to create a file. I created a file with 1738 unique records and then tried to load it and found that I would get anywhere from 1728-1733 records actually loaded so it seems this "dummy" file replicates the issue fairly accurately. 

Attached are two files: the first (starting with "QDESM") is the EBCDIC file itself. The second is a snippet of code used to create the IFileLayout to load the data. Note that this is not the entire class as I removed some extraneous entries that don't apply to this file but this still seems to replicate the issue just fine. But it should be enough to go on to more or less recreate what I am trying to do
[IFileLayout.txt](https://github.com/MistakenIdentity/WillPower/files/8928698/IFileLayout.txt)
[QDESM32022041301590000Y.txt](https://github.com/MistakenIdentity/WillPower/files/8928699/QDESM32022041301590000Y.txt)
.

**Comment by DarkAngel766** [comment]
Hi Andrew. This may be a matter of overdriving. The default maximum concurrency value is 256.  Can you reduce that to say, 32, and rerun your tests?  You can find it at FileParser.TaskManager.MaximumConcurrency

**Comment by andrew-steinkopf** [comment]
That didn't work either. I set it to 32 and the number of records loaded from the sample file was 1733 instead of 1739. I tried 1, 2, 5, 512, 1800, and int.maxvalue just to see what, if anything, would change. Setting it to 1 only let 1 record be loaded. 2 loaded 3 records. 5 loaded 5 records. The rest topped out at the 1729-1735 range again. So that doesn't appear to have any bearing on the underlying issue.
